### PR TITLE
Remove sklearn as a dependency by bringing in classification_report

### PR DESCRIPTION
**Pull request by Anthuang**
Before:
```bash
❯ du -h -d1 .venv/lib/python3.8/  
914M    .venv/lib/python3.8//site-packages
```

After:
```bash
❯ du -h -d1 .venv_test/lib/python3.8/
809M    .venv_test/lib/python3.8//site-packages
```

### CODE REVIEW

**Comment by thomasehuang** [comment]
Looks good so far. Make sure you run linting + test cases.

File: scalabel/eval/tagging.py [hunk 1]
```diff
@@ -17,6 +32,503 @@
...
+            return np.array([table[v] for v in values])
+        except KeyError as e:
+            raise ValueError(f"y contains previously unseen labels: {str(e)}")
+    else:
+        return np.searchsorted(uniques, values)
+
+
+def _unique_np(
+    values: Sequence[np.int32],
+    return_inverse: bool = False,
```
Comment (thomasehuang) [inline scalabel/eval/tagging.py:176]: This is always False, remove corresponding code

File: scalabel/eval/tagging.py [hunk 2]
```diff
@@ -17,6 +32,503 @@
...
+    return ret[0] if len(ret) == 1 else ret
+
+
+def _get_counts(values: NDArrayI32, uniques: Sequence[np.int32]):
+    """Get the count of each of the `uniques` in `values`.
+
+    The counts will use the order passed in by `uniques`. For non-object dtypes,
+    `uniques` is assumed to be sorted and `np.nan` is at the end.
+    """
+    if values.dtype.kind in "OU":
```
Comment (thomasehuang) [inline scalabel/eval/tagging.py:228]: Pretty sure this is never used, can you check?

File: scalabel/eval/tagging.py [hunk 3]
```diff
@@ -17,6 +32,503 @@
...
+            f"strings or numbers. Got {types}"
+        )
+    ret = (uniques,)
+
+    return ret[0] if len(ret) == 1 else ret
+
+
+def _unique(values: NDArrayI32) -> Sequence[np.int32]:
+    """Helper function to find unique values with support for python objects."""
+    if values.dtype == object:
```
Comment (thomasehuang) [inline scalabel/eval/tagging.py:273]: Pretty sure the only dtype our values have is float / int

File: scalabel/eval/tagging.py [hunk 4]
```diff
@@ -17,6 +32,503 @@
...
+    if return_inverse:
+        ret += (inverse,)
+
+    if return_counts:
+        ret += (counts,)
+
+    return ret[0] if len(ret) == 1 else ret
+
+
+def _get_counts(values: NDArrayI32, uniques: Sequence[np.int32]):
```
Comment (thomasehuang) [inline scalabel/eval/tagging.py:222]: Doesn't seem like this is called?

File: scalabel/eval/tagging.py [hunk 5]
```diff
@@ -17,6 +32,434 @@
...
+        return np.ravel(y)
+
+    raise ValueError(
+        "y should be a 1d array, got an array of shape {} instead.".format(
+            shape
+        )
+    )
+
+
+def _num_samples(x: NDArrayI32) -> int:
```
Comment (thomasehuang) [inline scalabel/eval/tagging.py:51]: This shouldn't be necessary, you can replace it with len(y) at where its being called

File: scalabel/eval/tagging.py [hunk 6]
```diff
@@ -17,6 +32,434 @@
...
+    # zero, and zero_division=1. In all other case, 0
+    if np.isposinf(beta):
+        f_score = recall
+    else:
+        denom = beta2 * precision + recall
+
+        denom[denom == 0.0] = 1  # avoid division by 0
+        f_score = (1 + beta2) * precision * recall / denom
+
+    if average == "weighted":
```
Comment (thomasehuang) [inline scalabel/eval/tagging.py:382]: Is this ever used? I dint remember using this

File: scalabel/eval/tagging.py [hunk 7]
```diff
@@ -17,6 +32,434 @@
...
+        out = np.diff(x.indptr)
+        return out * sample_weight
+    elif axis == 0:
+        weights = np.repeat(sample_weight, np.diff(x.indptr))
+        return np.bincount(x.indices, minlength=x.shape[1], weights=weights)
+    else:
+        raise ValueError("Unsupported axis: {0}".format(axis))
+
+
+def _multilabel_confusion_matrix(
```
Comment (thomasehuang) [inline scalabel/eval/tagging.py:273]: Rename function to `confusion_matrix`

**Review by thomasehuang — CHANGES_REQUESTED** [review]

File: scalabel/eval/tagging.py (hunk 6 above)
Comment (Anthuang) [reply scalabel/eval/tagging.py:382]: I'll take it out

File: scalabel/eval/tagging.py [hunk 8]
```diff
@@ -1,22 +1,252 @@
...
+    """Helper function to find unique values, accounts for nans."""
+    uniques: NDArrayI32 = np.unique(values)
+
+    # np.unique will have duplicate missing values at the end of `uniques`
+    # here we clip the nans and remove it from uniques
+    if (
+        uniques.size
+        and isinstance(uniques[-1], numbers.Real)
+        and math.isnan(uniques[-1])
+    ):
```
Comment (thomasehuang) [inline scalabel/eval/tagging.py:55]: Check if this is run ever

File: scalabel/eval/tagging.py [hunk 9]
```diff
@@ -1,22 +1,252 @@
...
+    y_true: NDArrayI32, y_pred: NDArrayI32, labels: NDArrayI32
+) -> NDArrayI32:
+    """Compute a confusion matrix for each class or sample."""
+    present_labels = _unique_labels(y_true, y_pred)
+    n_labels = labels.size
+    labels = np.hstack(
+        [labels, np.setdiff1d(present_labels, labels, assume_unique=True)]
+    )
+
+    if y_true.ndim == 1:
```
Comment (thomasehuang) [inline scalabel/eval/tagging.py:80]: No need for this if statement

File: scalabel/eval/tagging.py [hunk 10]
```diff
@@ -1,22 +1,252 @@
...
+
+    if not np.any(mask):
+        return result
+
+    # address zero division by setting 0s to 1s
+    result[mask] = 0
+    return result
+
+
+def _precision_recall_fscore(
```
Comment (thomasehuang) [inline scalabel/eval/tagging.py:137]: Let's just not use underscore prefix for any functions

**Review by thomasehuang — APPROVED** [review]
LGTM!

**Pull request merged**